### PR TITLE
test(openfeature): add null targeting key test coverage

### PR DESCRIPTION
**Pull request by leoromanovsky**
**What does this PR do?**

Adds test coverage for OpenFeature spec OF.2 (optional targeting key) to prove libdatadog handles nil targeting key correctly via the Ruby FFI layer.

**Motivation:**

The OpenFeature spec (Requirement 3.1.1, OF.2) defines targeting key as optional. The correct behavior (matching libdatadog's reference implementation) is:

| | Static (no rules, no shards) | Split (shards) | Targeted (rules) |
|---|---|---|---|
| **`""` empty string** | Returns value | Hashes `""`, returns value | Evaluates rules with `id=""`, returns value |
| **`null`/missing** | Returns value | `TARGETING_KEY_MISSING` error, returns default | Evaluates rules (no `id` fallback), returns value or falls through |

Ruby's evaluator delegates to libdatadog which already handles this correctly — the C extension passes `NULL` when `targeting_key` is absent from the context hash. This PR adds explicit test coverage proving the behavior.

**Change log entry**

No customer-visible change. Test-only — adds coverage for optional targeting key behavior that already works correctly.

**Additional Notes:**

- Tests placed in `spec/datadog/open_feature/` so `@DataDog/feature-flagging-and-experimentation-sdk` is auto-requested via CODEOWNERS
- Tests call `Configuration#get_assignment` directly to exercise the Rust FFI layer without mocks
- Used partial shard range (0-4999 of 10000) instead of full range to avoid libdatadog's insignificant shard optimization
- No logic changes to Ruby code

**Question for reviewers:** The existing `evaluation_engine_spec.rb` mocks `NativeEvaluator`, so there's no integration spec testing the full OpenFeature provider → native evaluator → libdatadog path with real flag configs. Should we add one?

**How to test the change?**

```bash
bundle exec rspec spec/datadog/open_feature/null_targeting_key_spec.rb
```

3 scenarios:
1. Static flag + nil targeting key → returns value with `STATIC` reason
2. Sharded flag + nil targeting key → returns `TARGETING_KEY_MISSING` error
3. Rule-match flag + nil targeting key → returns value with `TARGETING_MATCH` reason

## Next Steps

- Add `test-case-null-targeting-key.json` to `ffe-system-test-data` shared fixtures (separate fixture centralization project)
- Wire the fixture loop spec to cover null targeting key via JSON fixtures
- System test coverage for null targeting key across all tracers

### CODE REVIEW

File: spec/datadog/open_feature/null_targeting_key_spec.rb [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+# Tests for OpenFeature spec OF.2: optional targeting key
+# See: https://openfeature.dev/specification/sections/evaluation-context#requirement-311
+#
+# When targeting key is missing (nil), SDKs must still attempt evaluation:
+# - Static flags (no shards): return value normally
+# - Sharded flags: return TARGETING_KEY_MISSING error
+# - Rule-match flags (no shards): return value if rule matches on non-id attribute
+
+RSpec.describe 'OpenFeature OF.2: Optional Targeting Key' do
```
Comment (dd-oleksii) [inline spec/datadog/open_feature/null_targeting_key_spec.rb]: minor: OF.2 is our internal SDK spec identifier, not openfeature

**Review by dd-oleksii — APPROVED** [review]

**Review by Strech — APPROVED** [review]

File: spec/datadog/open_feature/null_targeting_key_spec.rb [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+#
+# When targeting key is missing (nil), SDKs must still attempt evaluation:
+# - Static flags (no shards): return value normally
+# - Sharded flags: return TARGETING_KEY_MISSING error
+# - Rule-match flags (no shards): return value if rule matches on non-id attribute
+
+RSpec.describe 'Datadog Provider OF.2: Optional Targeting Key' do
+  before do
+    skip 'Requires native libdatadog extension' unless Datadog::Core::FeatureFlags::Configuration.method(:new).arity == 1
+  end
```
Comment (p-datadog) [inline spec/datadog/open_feature/null_targeting_key_spec.rb:17]: question (non-blocking): These tests call the native C extension directly (`Configuration.new` → `get_assignment`), but `spec:open_feature` doesn't have a `.enhance([:compile])` dependency in the Rakefile — unlike `spec:core_with_libdatadog_api` (Rakefile:260) which does.

Right now the extension is available because the batch seed happens to place a compile-triggering task (like `core_with_libdatadog_api`) in the same CI job, and it runs first. A different seed could put `open_feature` in a batch without one, and the skip guard here would silently skip all three tests.

Would it make sense to add something like this to the Rakefile?

```ruby
Rake::Task['spec:open_feature'].enhance(["compile:libdatadog_api.#{RUBY_VERSION[/\d+.\d+/]}_#{RUBY_PLATFORM}"])
```

**Review by p-datadog — COMMENTED** [review]
One question below.

**Comment by p-datadog** [comment]
@codex review

**Comment by ivoanjo** [comment]
Hey, is this something we still want to merge?

**Comment by leoromanovsky** [comment]
> Hey, is this something we still want to merge?

Yes please; I got tired of waiting for CI to pass and forgot about it.

File: spec/datadog/open_feature/null_targeting_key_spec.rb [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+            "allocations": [{"key": "rule-alloc", "rules": [{"conditions": [{"attribute": "email", "operator": "MATCHES", "value": "@example\\\\.com"}]}], "splits": [{"variationKey": "matched", "shards": []}], "doLog": false}]
+          }
+        }
+      }
+    JSON
+  end
+
+  let(:config) { Datadog::Core::FeatureFlags::Configuration.new(flags_json) }
+
+  context 'with nil targeting key' do
```
Comment (typotter) [inline spec/datadog/open_feature/null_targeting_key_spec.rb]: nit: this is a missing targeting key as opposed to `nil`

File: spec/datadog/open_feature/null_targeting_key_spec.rb [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+# See: https://openfeature.dev/specification/sections/evaluation-context#requirement-311
+#
+# When targeting key is missing (nil), SDKs must still attempt evaluation:
+# - Static flags (no shards): return value normally
+# - Sharded flags: return TARGETING_KEY_MISSING error
+# - Rule-match flags (no shards): return value if rule matches on non-id attribute
+
+RSpec.describe 'Datadog Provider OF.2: Optional Targeting Key' do
+  before do
+    skip 'Requires native libdatadog extension' unless Datadog::Core::FeatureFlags::Configuration.method(:new).arity == 1
```
Comment (typotter) [inline spec/datadog/open_feature/null_targeting_key_spec.rb]: 🤖 claude is suggesting scoping the check to  
`defined?(Datadog::Core::FeatureFlags::NativeEvaluator)` here because this test will silently stop skipping if the C extension signature changes.

**Review by typotter — APPROVED** [review]

**Comment by p-datadog** [comment]
I took the liberty to have my LLM apply DI testing rules to this PR so that the added tests run in CI.

**Review by vpellan — APPROVED** [review]

**Comment by vpellan** [comment]
Merging this PR as it @leoromanovsky gave his green light in a previous comment
